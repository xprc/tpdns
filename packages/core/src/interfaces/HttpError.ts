export interface HttpError extends Record<string, any> {
    message: string;
    statusCode: number;
}
