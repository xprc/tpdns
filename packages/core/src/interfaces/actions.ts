export type Action = "create" | "edit" | "list" | "show" | "clone";

export type RouteAction = Exclude<Action, "list"> | undefined;

export type RedirectAction =
    | Extract<Action, "list" | "show" | "edit" | "create">
    | false;

/**
 * @deprecated use RedirectAction type instead
 */
export type RedirectionTypes = RedirectAction;

export type FormAction = Extract<Action, "create" | "edit" | "clone">;

export type ActionWithPage = Extract<Action, "show" | "create" | "edit">;
