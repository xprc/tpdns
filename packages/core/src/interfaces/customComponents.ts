import React, { ReactNode } from "react";

export type TitleProps = {
    collapsed: boolean;
};

export type LayoutProps = {
    Sider?: React.FC;
    Header?: React.FC;
    Title?: React.FC<TitleProps>;
    Footer?: React.FC;
    OffLayoutArea?: React.FC;
    dashboard?: boolean;
    children?: ReactNode;
};

export type DashboardPageProps<TCrudData = any> = {
    initialData?: TCrudData;
} & Record<any, any>;
