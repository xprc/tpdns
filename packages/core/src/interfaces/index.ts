// contexts
export * from "../contexts/data/IDataContext";
export * from "../contexts/live/ILiveContext";
export * from "../contexts/auth/IAuthContext";
export * from "../contexts/refine/IRefineContext";
export * from "../contexts/translation/ITranslationContext";
export * from "../contexts/undoableQueue/IUndoableQueueContext";
export * from "../contexts/resource/IResourceContext";
export * from "../contexts/unsavedWarn/IUnsavedWarnContext";
export * from "../contexts/router/IRouterContext";
export * from "../contexts/accessControl/IAccessControlContext";
export * from "../contexts/notification/INotificationContext";
export * from "../contexts/auditLog/IAuditLogContext";

export * from "../components/pages/login";

// actions
export * from "./actions";

// notification
export * from "./notification";

// mutationMode
export * from "./mutationMode";

// mutationMode
export * from "./HttpError";

// custom components
export * from "./customComponents";

// resourceRouterParams
export * from "./resourceRouterParams";

// resourceErrorRouterParams
export * from "./resourceErrorRouterParams";

// mapData
export * from "./mapDataFn";

// successErrorNotification
export * from "./successErrorNotification";

//metaData
export * from "./metaData";

//queryKeys
export * from "./queryKey";

//menuItem
export * from "./menuItem";

//metaData
export * from "./live";

//auditLog
export * from "./auditLog";

export type BaseKey = string | number;
export type BaseRecord = {
    id?: BaseKey;
    // eslint-disable-next-line @typescript-eslint/no-explicit-any
    [key: string]: any;
};
export interface Option {
    label: string;
    value: string;
}

export * from "./auth";
