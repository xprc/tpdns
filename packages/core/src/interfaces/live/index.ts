export * from "./LiveEvent";
export * from "./LiveModeProps";
