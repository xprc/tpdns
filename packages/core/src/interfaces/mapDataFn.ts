export interface MapDataFn<TItem, TVariables> {
    (item: TItem, index?: number, items?: TItem[]): TVariables;
}
