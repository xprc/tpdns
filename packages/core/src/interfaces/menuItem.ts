import { IResourceItem } from "../contexts/resource/IResourceContext";

export type ITreeMenu = IResourceItem & {
    children: ITreeMenu[];
};

export type IMenuItem = IResourceItem & {
    key: string;
    route: string;
};
