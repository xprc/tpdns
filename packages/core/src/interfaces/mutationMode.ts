import {
    BaseRecord,
    GetListResponse,
    GetOneResponse,
    IQueryKeys,
} from "../interfaces";
import { QueryKey } from "@tanstack/react-query";

export type MutationMode = "pessimistic" | "optimistic" | "undoable";

export type QueryResponse<T = BaseRecord> =
    | GetListResponse<T>
    | GetOneResponse<T>;

export type PreviousQuery<TData> = [QueryKey, TData | unknown];

export type PrevContext<TData> = {
    previousQueries: PreviousQuery<TData>[];
    queryKey: IQueryKeys;
};

export type Context = {
    previousQueries: ContextQuery[];
};

export type ContextQuery<T = BaseRecord> = {
    query: QueryResponse<T>;
    queryKey: QueryKey;
};
