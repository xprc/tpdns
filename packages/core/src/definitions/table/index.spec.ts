import {
    stringifyTableParams,
    parseTableParams,
    parseTableParamsFromQuery,
    unionFilters,
    compareFilters,
    compareSorters,
    unionSorters,
    getDefaultSortOrder,
    getDefaultFilter,
} from "./";
import { CrudSorting, CrudFilters } from "../../interfaces";

describe("definitions/table", () => {
    it("getDefaultSortOrder", () => {
        const sorter: CrudSorting = [
            {
                field: "title",
                order: "asc",
            },
            {
                field: "view",
                order: "desc",
            },
        ];

        expect(getDefaultSortOrder("title", sorter)).toEqual("asc");
        expect(getDefaultSortOrder("view", sorter)).toEqual("desc");
    });

    it("getDefaultFilter", () => {
        const filters: CrudFilters = [
            {
                field: "title",
                operator: "contains",
                value: "test",
            },
        ];
        expect(getDefaultFilter("title", filters, "contains")).toEqual("test");
    });

    it("getDefaultFilter empty array", () => {
        const filters: CrudFilters = [
            {
                field: "title",
                operator: "contains",
                value: undefined,
            },
        ];
        expect(getDefaultFilter("title", filters, "contains")).toEqual([]);
    });

    it("getDefaultFilter default operator", () => {
        const filters: CrudFilters = [
            {
                field: "title",
                operator: "eq",
                value: "test",
            },
        ];
        expect(getDefaultFilter("title", filters)).toEqual("test");
    });

    it("stringify table params correctly", async () => {
        const pagination = {
            current: 1,
            pageSize: 10,
        };

        const sorter: CrudSorting = [
            {
                field: "id",
                order: "desc",
            },
            {
                field: "title",
                order: "desc",
            },
        ];

        const filters: CrudFilters = [
            {
                field: "categoryId",
                operator: "in",
                value: [1, 2],
            },
        ];

        const userDefinedQueryParam = {
            foo: "bar",
        };

        const url = stringifyTableParams({
            pagination,
            sorter,
            filters,
            ...userDefinedQueryParam,
        });
        expect(url).toMatchSnapshot();
    });

    it("stringify table single sort params correctly", async () => {
        const pagination = {
            current: 1,
            pageSize: 10,
        };

        const sorter: CrudSorting = [{ field: "id", order: "desc" }];
        const filters: CrudFilters = [
            {
                field: "categoryId",
                operator: "in",
                value: [1, 2],
            },
        ];

        const url = stringifyTableParams({ pagination, sorter, filters });
        expect(url).toMatchSnapshot();
    });

    it("parse table params with single sorter correctly", async () => {
        const url =
            "?current=1&pageSize=10&sorter[0][field]=id&sorter[0][order]=desc&filters[0][operator]=in&filters[0][field]=categoryId&filters[0][value][0]=1&filters[0][value][1]=2";

        const { parsedCurrent, parsedPageSize, parsedSorter, parsedFilters } =
            parseTableParams(url);

        expect(parsedCurrent).toBe(1);
        expect(parsedPageSize).toBe(10);
        expect(parsedSorter).toStrictEqual([{ field: "id", order: "desc" }]);
        expect(parsedFilters).toStrictEqual([
            { field: "categoryId", operator: "in", value: ["1", "2"] },
        ]);
    });

    it("parse table params with advanced query object", async () => {
        const query = {
            current: 1,
            pageSize: 10,
            sorter: [
                { field: "id", order: "asc" },
                { field: "firstName", order: "desc" },
            ],
            filters: [
                { field: "id", operator: "eq", value: "1" },
                {
                    operator: "or",
                    value: [
                        {
                            field: "age",
                            operator: "lt",
                            value: "18",
                        },
                        {
                            field: "age",
                            operator: "gt",
                            value: "20",
                        },
                    ],
                },
            ],
        };

        const { parsedCurrent, parsedPageSize, parsedSorter, parsedFilters } =
            parseTableParamsFromQuery(query);

        expect(parsedCurrent).toBe(1);
        expect(parsedPageSize).toBe(10);
        expect(parsedSorter).toStrictEqual([
            { field: "id", order: "asc" },
            { field: "firstName", order: "desc" },
        ]);
        expect(parsedFilters).toStrictEqual([
            { field: "id", operator: "eq", value: "1" },
            {
                operator: "or",
                value: [
                    {
                        field: "age",
                        operator: "lt",
                        value: "18",
                    },
                    {
                        field: "age",
                        operator: "gt",
                        value: "20",
                    },
                ],
            },
        ]);
    });

    it("unionFilters should override same filters", () => {
        expect(
            unionFilters(
                [
                    {
                        field: "foo",
                        operator: "in",
                        value: "permenant",
                    },
                ],
                [
                    {
                        field: "foo",
                        operator: "in",
                        value: "crud",
                    },
                    {
                        field: "bar",
                        operator: "in",
                        value: "crud",
                    },
                ],
            ),
        ).toMatchInlineSnapshot(`
            Array [
              Object {
                "field": "foo",
                "operator": "in",
                "value": "permenant",
              },
              Object {
                "field": "bar",
                "operator": "in",
                "value": "crud",
              },
            ]
        `);
    });

    it("unionFilters should override even when filter value is null but should not keep it in the end result", () => {
        expect(
            unionFilters(
                [],
                [
                    {
                        field: "foo",
                        operator: "in",
                        value: null,
                    },
                    {
                        field: "bar",
                        operator: "in",
                        value: undefined,
                    },
                    {
                        field: "baz",
                        operator: "in",
                        value: "prev",
                    },
                ],
            ),
        ).toMatchInlineSnapshot(`
            Array [
              Object {
                "field": "baz",
                "operator": "in",
                "value": "prev",
              },
            ]
        `);
    });

    it("compareFilters filters are the same if their field and operator are the same", () => {
        expect(
            compareFilters(
                {
                    field: "foo",
                    operator: "in",
                    value: "left",
                },
                {
                    field: "foo",
                    operator: "in",
                    value: "right",
                },
            ),
        ).toBe(true);

        expect(
            compareFilters(
                {
                    field: "foo",
                    operator: "in",
                    value: "test",
                },
                {
                    field: "foo",
                    operator: "contains",
                    value: "test",
                },
            ),
        ).toBe(false);
    });

    it("compareFilters return correct result when `or` is used", () => {
        expect(
            compareFilters(
                {
                    operator: "or",
                    value: [
                        {
                            field: "name",
                            operator: "eq",
                            value: "test",
                        },
                    ],
                },
                {
                    operator: "or",
                    value: [
                        {
                            field: "created_at",
                            operator: "gte",
                            value: "2022-01-01",
                        },
                    ],
                },
            ),
        ).toBe(true);

        expect(
            compareFilters(
                {
                    operator: "or",
                    value: [
                        {
                            field: "name",
                            operator: "eq",
                            value: "test",
                        },
                    ],
                },
                {
                    field: "created_at",
                    operator: "gte",
                    value: "2022-01-01",
                },
            ),
        ).toBe(false);
    });

    it("unionFilters should override `or` filter", () => {
        const union = unionFilters(
            // permanent filters
            [],
            // new filters
            [
                {
                    field: "other-field",
                    operator: "in",
                    value: "crud",
                },
                {
                    operator: "or",
                    value: [
                        {
                            field: "created_at",
                            operator: "contains",
                            value: "2022",
                        },
                    ],
                },
            ],
            // prev filters
            [
                {
                    operator: "or",
                    value: [
                        {
                            field: "name",
                            operator: "eq",
                            value: "test",
                        },
                    ],
                },
            ],
        );

        // does not include previous `or`
        expect(union).not.toEqual(
            expect.arrayContaining([
                expect.objectContaining({
                    operator: "or",
                    value: expect.arrayContaining([
                        expect.objectContaining({
                            field: "name",
                            operator: "eq",
                            value: "test",
                        }),
                    ]),
                }),
            ]),
        );

        // includes new `or` and new filters
        expect(union).toMatchObject([
            {
                field: "other-field",
                operator: "in",
                value: "crud",
            },
            {
                operator: "or",
                value: [
                    {
                        field: "created_at",
                        operator: "contains",
                        value: "2022",
                    },
                ],
            },
        ]);
    });

    it("unionFilters should remove `or` filter if value is empty array", () => {
        const union = unionFilters(
            // permanent filters
            [],
            // new filters
            [
                {
                    field: "other-field",
                    operator: "in",
                    value: "crud",
                },
                {
                    operator: "or",
                    value: [],
                },
            ],
            // prev filters
            [
                {
                    operator: "or",
                    value: [
                        {
                            field: "name",
                            operator: "eq",
                            value: "test",
                        },
                    ],
                },
            ],
        );

        expect(union).not.toEqual(
            expect.arrayContaining([
                expect.objectContaining({
                    operator: "or",
                }),
            ]),
        );
    });

    it("unionFilters should keep `or` filter if it's untouched", () => {
        const union = unionFilters(
            // permanent filters
            [],
            // new filters
            [
                {
                    field: "other-field",
                    operator: "in",
                    value: "crud",
                },
            ],
            // prev filters
            [
                {
                    operator: "or",
                    value: [
                        {
                            field: "name",
                            operator: "eq",
                            value: "test",
                        },
                    ],
                },
            ],
        );

        expect(union).toEqual(
            expect.arrayContaining([
                expect.objectContaining({
                    operator: "or",
                }),
            ]),
        );
    });

    it("unionSorters should override same sorter", () => {
        expect(
            unionSorters(
                [
                    {
                        field: "foo",
                        order: "asc",
                    },
                ],
                [
                    {
                        field: "foo",
                        order: "asc",
                    },
                    {
                        field: "bar",
                        order: "asc",
                    },
                ],
            ),
        ).toMatchInlineSnapshot(`
            Array [
              Object {
                "field": "foo",
                "order": "asc",
              },
              Object {
                "field": "bar",
                "order": "asc",
              },
            ]
        `);
    });

    it("unionSorters should override even when sorter value is null but should not keep it in the end result", () => {
        expect(
            unionSorters(
                [],
                [
                    {
                        field: "bar",
                        order: "asc",
                    },
                    {
                        field: "foo",
                        order: "asc",
                    },
                ],
            ),
        ).toMatchInlineSnapshot(`
            Array [
              Object {
                "field": "bar",
                "order": "asc",
              },
              Object {
                "field": "foo",
                "order": "asc",
              },
            ]
        `);
    });

    it("compareSorters sorters are the same if their field are the same", () => {
        expect(
            compareSorters(
                {
                    field: "foo",
                    order: "asc",
                },
                {
                    field: "foo",
                    order: "asc",
                },
            ),
        ).toBe(true);

        expect(
            compareSorters(
                {
                    field: "foo",
                    order: "asc",
                },
                {
                    field: "foo",
                    order: "desc",
                },
            ),
        ).toBe(true);
    });
});
