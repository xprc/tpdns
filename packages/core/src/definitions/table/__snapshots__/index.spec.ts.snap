// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`definitions/table stringify table params correctly 1`] = `"foo=bar&current=1&pageSize=10&sorter[0][field]=id&sorter[0][order]=desc&sorter[1][field]=title&sorter[1][order]=desc&filters[0][field]=categoryId&filters[0][operator]=in&filters[0][value][0]=1&filters[0][value][1]=2"`;

exports[`definitions/table stringify table single sort params correctly 1`] = `"current=1&pageSize=10&sorter[0][field]=id&sorter[0][order]=desc&filters[0][field]=categoryId&filters[0][operator]=in&filters[0][value][0]=1&filters[0][value][1]=2"`;
